Frameworks/CoreGraphics/CGImage.mm

```diff
@@ -116,271 +51,93 @@ COREGRAPHICS_EXPORT void CGImageAddDestructionListener(CGImageDestructionListene
-    surfaceInfo.width = rect.size.width;
-    surfaceInfo.height = rect.size.height;
-
-    assert(surfaceInfo.surfaceData == NULL);
+CGImageRef CGImageCreate(size_t width,
+                         size_t height,
+                         size_t bitsPerComponent,
+                         size_t bitsPerPixel,
+                         size_t bytesPerRow,
+                         CGColorSpaceRef colorSpace,
+                         CGBitmapInfo bitmapInfo,
+                         CGDataProviderRef provider,
+                         const float* decode,
+                         bool shouldInterpolate,
+                         CGColorRenderingIntent intent) {
+    RETURN_NULL_IF(((provider == nullptr) || ![(NSObject*)provider isKindOfClass:[NSData class]]) || (colorSpace == nullptr));
 
-    CGImageRef newImage = new CGBitmapImage(surfaceInfo);
+    NSData* dataProvider = (__bridge NSData*)provider;
 
-    int startX = (int)rect.origin.x;
-    int startY = (int)rect.origin.y;
-    int sizeX = (int)rect.size.width;
-    int sizeY = (int)rect.size.height;
+    unsigned char* data = (unsigned char*)[dataProvider bytes];
 
-    BYTE* srcIn =
-        ((BYTE*)ref->Backing()->LockImageData()) + startY * ref->Backing()->BytesPerRow() + startX * ref->Backing()->BytesPerPixel();
-    BYTE* destOut = (BYTE*)newImage->Backing()->LockImageData();
+    ComPtr<IWICBitmap> image;
+    ComPtr<IWICImagingFactory> imageFactory = _GetWICFactory();
 
-    for (int curY = 0; curY < sizeY; curY++) {
-        memmove(destOut, srcIn, newImage->Backing()->BytesPerRow());
+    REFGUID pixelFormat = _CGImageGetWICPixelFormat(bitsPerComponent, bitsPerPixel, colorSpace, bitmapInfo);
 
-        srcIn += ref->Backing()->BytesPerRow();
-        destOut += newImage->Backing()->BytesPerRow();
-    }
+    HRESULT status = imageFactory->CreateBitmapFromMemory(width, height, pixelFormat, bytesPerRow, height * bytesPerRow, data, &image);
 
-    ref->Backing()->ReleaseImageData();
-    newImage->Backing()->ReleaseImageData();
+    CGImageRef imageRef = __CGImage::CreateInstance();
+    imageRef->SetImageSource(image).SetColorSpace(colorSpace).SetRenderingIntent(intent).SetInterpolate(shouldInterpolate);
 
-    return (CGImageRef)newImage;
+    return imageRef;
 }
 
 /**
  @Status Interoperable
- @Notes Doesn't support copy-on-write semantics - returns an unlinked copy of the source
-        image.
 */
-CGImageRef CGImageCreateCopy(CGImageRef ref) {
-    if (!ref)
-        return nullptr;
-
-    __CGSurfaceInfo surfaceInfo;
-    ref->Backing()->GetSurfaceInfoWithoutPixelPtr(&surfaceInfo);
-
-    assert(surfaceInfo.surfaceData == NULL);
-
-    CGImageRef newImage = new CGBitmapImage(surfaceInfo);
-
-    int startX = 0;
-    int startY = 0;
-    int sizeX = ref->Backing()->Width();
-    int sizeY = ref->Backing()->Height();
-
-    if (startY < 0) {
-        startY = 0;
-    }
+CGImageRef CGImageCreateWithImageInRect(CGImageRef ref, CGRect rect) {
+    RETURN_NULL_IF(!ref);
```

Comment:
this is a valid keyword

---
In reply to: [85252277](https://github.com/Microsoft/WinObjC/pull/1185#discussion_r85252277) [](ancestors = 85252277)